packages/kbn-discover-utils/src/components/app_menu/app_menu_registry.ts

```diff
@@ -0,0 +1,193 @@
+    this.customSubmenuItemsBySubmenuId = new Map();
+  }
+
+  public isActionRegistered(appMenuItemId: string) {
+    return (
+      this.appMenuItems.some((item) => {
+        if (item.id === appMenuItemId) {
+          return true;
+        }
+        if (isAppMenuActionSubmenu(item)) {
+          return item.actions.some((submenuItem) => submenuItem.id === appMenuItemId);
+        }
+        return false;
+      }) ||
+      [...this.customSubmenuItemsBySubmenuId.values()].some((submenuItems) =>
+        submenuItems.some((item) => item.id === appMenuItemId)
+      )
+    );
+  }
+
+  /**
+   * Register a custom action to the app menu. It can be a simple action or a submenu with more actions and horizontal rules.
+   * Note: Only 2 top level custom actions are allowed to be rendered in the app menu. The rest will be ignored.
+   * A custom action can also open a flyout or a modal. For that, return your custom react node from action's `onClick` event and call `onFinishAction` when you're done.
+   * @param appMenuItem
+   */
+  public registerCustomAction(appMenuItem: AppMenuItemCustom) {
+    this.appMenuItems = [
+      ...this.appMenuItems.filter(
+        // prevent duplicates
+        (item) => !(item.id === appMenuItem.id && item.type === AppMenuActionType.custom)
+      ),
+      appMenuItem,
+    ];
+  }
+
+  /**
+   * Register a custom action under a submenu. It can be an action or a horizontal rule.
+   * Any number of submenu actions can be registered and rendered.
+   * You can also extend an existing submenu with more actions. For example, AppMenuActionType.alerts.
+   * `order` property is optional and can be used to control the order of actions in the submenu.
+   * @param submenuId
+   * @param appMenuItem
+   */
+  public registerCustomActionUnderSubmenu(submenuId: string, appMenuItem: AppMenuActionCustom) {
+    this.customSubmenuItemsBySubmenuId.set(submenuId, [
+      ...(this.customSubmenuItemsBySubmenuId.get(submenuId) ?? []).filter(
+        // prevent duplicates
+        (item) => !(item.id === appMenuItem.id && item.type === AppMenuActionType.custom)
+      ),
+      appMenuItem,
+    ]);
+  }
+
+  private getSortedItemsForType(type: AppMenuActionType) {
+    let actions = this.appMenuItems.filter((item) => item.type === type);
+
+    if (type === AppMenuActionType.custom && actions.length > AppMenuRegistry.CUSTOM_ITEMS_LIMIT) {
+      // apply the limitation on how many custom items can be shown
+      actions = actions.slice(0, AppMenuRegistry.CUSTOM_ITEMS_LIMIT);
+    }
+
+    // enrich submenus with custom actions
+    if (type === AppMenuActionType.secondary || type === AppMenuActionType.custom) {
+      [...this.customSubmenuItemsBySubmenuId.entries()].forEach(([submenuId, customActions]) => {
+        const submenuParentItem = actions.find((item) => item.id === submenuId);
+        if (submenuParentItem && isAppMenuActionSubmenu(submenuParentItem)) {
+          submenuParentItem.actions.push(...customActions);
```

Comment:
Won't pushing to these arrays result in duplicate actions if `getSortedItems` is ever called more than once on the same instance? Or is the intent that `getSortedItems` should only ever be called once?

Reply:
Good catch! Fixed in https://github.com/elastic/kibana/pull/195448/commits/fecdaad3763cc43c8dd815d2c14f216e87b60945